sos/cleaner/__init__.py

```diff
@@ -0,0 +1,552 @@
+                    self.log_debug("Archive %s failed to compress: %s"
+                                   % (archive.archive_name, err))
+                    archive.report_msg("Failed to re-compress archive: %s"
+                                       % err)
+                    return
+
+            self.completed_reports.append(archive)
+            archive.report_msg("Obfuscation completed")
+
+        except Exception as err:
+            self.ui_log.info("Exception while processing %s: %s"
+                             % (report, err))
+
+    def prep_maps_from_archive(self, archive):
+        """Open specific files from an archive and try to load those values
+        into our mappings before iterating through the entire archive.
+
+        Positional arguments:
+
+            :param archive SoSObfuscationArchive:   An open archive object
+        """
+        for parser in self.parsers:
+            # this is a bit clunky, but we need to load this particular
+            # parser in a different way due to how hostnames are validated for
+            # obfuscation
+            if isinstance(parser, SoSHostnameParser):
+                hn_file = archive.get_file_path(parser.prep_map_file)
+                with open(hn_file, 'r') as host_file:
+                    hostname = host_file.readline().strip()
+                    parser.load_hostname_into_map(hostname)
+            else:
+                self.obfuscate_file(
+                    archive.get_file_path(parser.prep_map_file)
+                )
+
+    def obfuscate_mac_from_line(self, line):
+        """Parse a line for presence of a MAC address, and if any are found
+        perform in place substitutions
+        """
+        pass
+
+    def obfuscate_ip_from_line(self, line):
+        """Parse a line for presence of an IP address, and if any are found
+        perform in place substitutions
+        """
+        pass
+
+    def obfuscate_file(self, filename, short_name=None):
+        """Obfuscate and individual file, line by line.
+
+        Lines processed, even if no substitutions occur, are then written to a
+        temp file without our own tmpdir. Once the file has been completely
+        iterated through, if there have been substitutions then the temp file
+        overwrites the original file. If there are no substitutions, then the
+        original file is left in place.
+
+        Positional arguments:
+
+            :param filename str:        Filename relative to the extracted
+                                        archive root
+        """
+        if not filename:
+            # the requested file doesn't exist in the archive
+            return
+        self.log_debug("Obfuscating %s" % filename)
+        subs = 0
+        tfile = tempfile.NamedTemporaryFile(mode='w', dir=self.tmpdir)
+        with open(filename, 'r') as fname:
+            for line in fname:
+                if not line.strip() or line.startswith('#'):
```

Comment:
A detail so far, but IMHO comment lines needs to be obfuscated as well. Assume configs like:

    connectTo: hostname.domain
    # connectTo: another.host.domain

Reply:
> * `--domains` requires an extra argument? Do we expect it is often used? (I dont have a user experience with sos-cleaner, just curious if one specific pattern desires an extra cmdline argument)

--domains allows a user to specify domain names that don't fall under the system's FQDN, which is all we obfuscate by default - there's simply too many false positives generated from any regex for FQDNs to be able to cover all found domains without disrupting a large amount of other data.

> * Methods shall be re-ordered to declare auxiliary ones first - this would helped me for readability of code.

Sure

> * Should not we obfuscate also filenames (e.g. even plain `sosreport-<hostname>-<timestamp>..` or others inside the report, like cmdline based filenames (e.g. https://github.com/sosreport/sos/blob/master/sos/report/plugins/pulp.py#L128 generates a filename containing `self.messaging_cert_file` that does contain FQDN of the system; similarly some remote DBs hostnames or IPs can occur in other cmdline based filenames))

That shouldn't be too difficult, we'd just need to add an `obfuscate_string()` method to handle non-file-content with the parsers. Not sure about the archive name though...

> * `generate_checksum`: shall not be a common routine also for sosreport, inside `archive` class?

 I can rename it, sure

Reply:
Hmm, actually on second thought let's discuss this one a bit further.

Comments in config files has the potential to throw off the used subnet relationships in sosreports.

For example - suppose the system has an actually used subnet of 10.11.12.0/24, with a bunch of addresses in that subnet in use. Now suppose a config file has a line of `# for example 10.11.12.0/26` or a similarly styled smaller subnet than what is actually used on the system.

Now any addresses in the /24 subnet on the system that would mathematically fall into the /26 subnet will now be obfuscated to the network created for /26 network from the comment.

We do already have some form of this problem with some /proc files (`/proc/pid/net/fib_trie`), but this would exacerbate the issue.

Reply:
I see and agree this is a concern we should handle. But:

- how often could configs contain such subnet comment that cause such problems = compared to comments with a valid hostname / IP we should obfuscate? IMHO the "not cleaning comments" is a bigger issue.

- what if the `10.11.12.0/26` really occurs in a non-commented config? :)

Reply:
Fair points both, ack to dropping the comment filter